### PR TITLE
CardViewer: Correctly Decode typed text in <input>

### LINKED ISSUE #5944: [ACRA 2.9.6] AstractFlashcardViewer - Link Parsing Error - fails when typing percent sign in input
Should be an easy one:

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/4751c711-f2ae-4b29-abef-fde2e8f4e581

```
java.lang.IllegalArgumentException: Invalid % sequence at 0: %
at libcore.net.UriCodec.decode(UriCodec.java:169)
at java.net.URLDecoder.decode(URLDecoder.java:60)
at com.ichi2.anki.AbstractFlashcardViewer$13.decodeUrl(AbstractFlashcardViewer.java:1)
at com.ichi2.anki.AbstractFlashcardViewer$13.filterUrl(AbstractFlashcardViewer.java:11)
at com.ichi2.anki.AbstractFlashcardViewer$13.shouldOverrideUrlLoading(AbstractFlashcardViewer.java:3)
at org.chromium.android_webview.AwContentsClientBridge.shouldOverrideUrlLoading(PG:17)
at android.os.MessageQueue.nativePollOnce(Native Method)
at android.os.MessageQueue.next(MessageQueue.java:323)
at android.os.Looper.loop(Looper.java:143)
at android.app.ActivityThread.main(ActivityThread.java:7224)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
at android.content.ContextWrapper.startService(ContextWrapper.java:675)
```

### PR DESCRIPTION
**Pull request by david-allison-1**
## Purpose / Description
Previously, only some data was URL Encoded when sent from JS to Java. When an invalid sequence (a single `%`) was entered, then the app would crash due to the decode failure.

## Fixes
Fixes #5944

## Approach
We both fix this encoding in the JS, and if invalid data comes in for any reason, we show an error.

## How Has This Been Tested?

Unit Tested with data. No longer crashes my Android.

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5327610&bid=17437212)

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 1]
```diff
@@ -3168,4 +2910,300 @@ public static int getSignalFromUrl(String url) {
...
+        @Override
+        @SuppressWarnings("deprecation") // tracked as #5017 in github
+        public boolean shouldOverrideUrlLoading(WebView view, String url) {
+            return filterUrl(url);
+        }
+
+
+        // Filter any links using the custom "playsound" protocol defined in Sound.java.
+        // We play sounds through these links when a user taps the sound icon.
+        private boolean filterUrl(String url) {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:2948]: Codacy Issue found: [The method 'filterUrl(String)' has an NPath complexity of 22176](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5327610)

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java [hunk 2]
```diff
@@ -272,4 +285,40 @@ public void invalidEaseIsParsedFromSignal() {
         String url = "signal:answer_ease0";
         assertEquals(SIGNAL_NOOP, getSignalFromUrl(url));
     }
+
+    @Test
+    public void invalidEncodingDoesNotCrash() {
```
Comment (timrae) [inline AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerTest.java:290]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5327610)

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 3]
```diff
@@ -1494,265 +1494,7 @@ private WebView createWebView() {
         webView.setScrollbarFadingEnabled(true);
         Timber.d("Focusable = %s, Focusable in touch mode = %s", webView.isFocusable(), webView.isFocusableInTouchMode());
 
-        webView.setWebViewClient(new WebViewClient() {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1497]: Yeah, maybe having a 250+ line sub-class inline isn't the best idea... ;-)

**Comment by mikehardy** [comment]
This released as 2.10alpha65 as soon as google is done processing it

**Pull request merged**